bson/src/test/unit/util/JsonPoweredTestHelper.java

```diff
@@ -27,24 +27,66 @@
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collection;
+import java.util.Collections;
 import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 public final class JsonPoweredTestHelper {
+    private static final String JSON_FILE_NAME_SUFFIX = ".json";
 
     public static BsonDocument getTestDocument(final File file) throws IOException {
         return new BsonDocumentCodec().decode(new JsonReader(getFileAsString(file)), DecoderContext.builder().build());
     }
 
+    public static BsonDocument getTestDocument(final Path file) throws IOException {
+        return getTestDocument(file.toFile());
+    }
+
     public static BsonDocument getTestDocument(final String resourcePath) throws IOException, URISyntaxException {
         return getTestDocument(new File(JsonPoweredTestHelper.class.getResource(resourcePath).toURI()));
     }
 
-    public static List<File> getTestFiles(final String resourcePath) throws URISyntaxException {
-        List<File> files = new ArrayList<File>();
-        addFilesFromDirectory(new File(JsonPoweredTestHelper.class.getResource(resourcePath).toURI()), files);
-        return files;
+    /**
+     * @see #getTestFiles(String, Path...)
+     */
+    public static Collection<File> getTestFiles(final String resourcePath) throws URISyntaxException, IOException {
+        return getTestFiles(resourcePath, (Path[]) null)
+                .stream()
+                .map(Path::toFile)
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * @param resourcePath             The {@linkplain Class#getResource(String) name of a resource} representing
+     *                                 a directory with test files.
+     * @param excludedRelativeSubpaths Sub-paths relative to the {@code resourcePath} directory to exclude from the results.
+     * @return A collection with unspecified mutability and element order.
+     */
+    public static Collection<Path> getTestFiles(final String resourcePath, final Path... excludedRelativeSubpaths)
+            throws URISyntaxException, IOException {
+        URL resource = JsonPoweredTestHelper.class.getResource(resourcePath);
+        if (resource == null) {
+            return Collections.emptyList();
+        } else {
+            Path start = Paths.get(resource.toURI());
+            List<Path> excluded = excludedRelativeSubpaths == null ? Collections.emptyList() : Stream.of(excludedRelativeSubpaths)
+                    .map(start::resolve)
+                    .collect(Collectors.toList());
+            try (Stream<Path> files = Files.walk(start)) {
+                return files
+                        .filter(path -> !Files.isDirectory(path))
+                        .filter(path -> path.getNameCount() > 0 && path.getFileName().toString().endsWith(JSON_FILE_NAME_SUFFIX))
+                        .filter(path -> excluded.stream().noneMatch(path::startsWith))
```

Comment:
The efficient way to implement exclusion of sub-paths is to use the `Files.walkFileTree` method and return `SKIP_SUBTREE` from `FileVisitor.preVisitDirectory`. However, for our purposes here making the code shorter seems to be more important, than making it efficient. However, if someone thinks this is a problem, I can implement exclusion via `Files.walkFileTree`.

Reply:
This comment is not relevant anymore.